### PR TITLE
feat: expression support for retry condition

### PR DESCRIPTION
**Pull request by StarpTech**
## Summary by CodeRabbit

- **New Features**
  - Expression-driven, configurable retry policies for subgraph requests with sensible defaults; retries skip mutations and honor Retry-After. Transport clients now wire logging, circuit-breaker, and conditional tracing; OnRetry callbacks receive computed sleep durations.
- **Configuration**
  - Retry expression exposed in config (env/YAML) with a safe default and new algorithm/max-duration/interval env bindings.
- **Tests**
  - Extensive unit and integration coverage for retry expressions, Retry-After semantics, callbacks, and edge cases.
- **Chores**
  - Dependency upgraded to expr-lang v1.17.6.
- **Breaking Changes**
  - Public retry API expanded; call sites updated to new signature.

# Example configuration

```yaml
traffic_shaping:
  all:
    retry:
      expression: IsRetryableStatusCode() || IsConnectionError() || IsTimeout()
```

# Retry Strategy Decision Table

| Scenario              | Expression                                                          | Description                                           |
| --------------------- | ------------------------------------------------------------------- | ----------------------------------------------------- |
| **Conservative**      | `IsRetryableStatusCode()`                                           | Only retry standard HTTP errors (500,502,503,504) |
| **Balanced**          | `IsRetryableStatusCode() \|\| IsTimeout() \|\| IsConnectionError()` | Most transient errors - recommended default           |
| **Aggressive**        | `Is5xxError() \|\| IsTimeout() \|\| IsConnectionError()`            | Maximum availability, higher retry volume             |
| **Timeout Only**      | `IsTimeout()`                                                       | Only network/HTTP timeouts                            |
| **HTTP Read Timeout** | `IsHttpReadTimeout()`                                               | Only HTTP read timeouts                               |
| **Connection Only**   | `IsConnectionError()`                                               | Only network connectivity issues                      |
| **Custom Status**     | `statusCode == 503`                                                 | Specific HTTP status codes                            |
| **Error Text**        | `error contains "timeout"`                                          | Match error message text                              |

**Available Methods:** `IsTimeout()`, `IsConnectionError()`, `IsConnectionRefused()`, `IsConnectionReset()`, `Is5xxError()`, `IsRetryableStatusCode()`, `IsHttpReadTimeout()`  
**Available Fields:** `statusCode` (int), `error` (string)

**Default:** `IsRetryableStatusCode() || IsConnectionError() || IsTimeout()`

**Note:** Mutations are never retried regardless of this expression. EOF errors are always retried.

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: router/internal/expr/retry_context.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
+package expr
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"syscall"
+)
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/internal/expr/retry_context.go:10]: Most cases should be handled by netErr.Timeout() anyway

File: router/internal/expr/retry_context.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	switch ctx.StatusCode {
+	case http.StatusInternalServerError,
+		http.StatusBadGateway,
+		http.StatusServiceUnavailable,
+		http.StatusGatewayTimeout:
+		return true
+	default:
+		return false
+	}
+}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/internal/expr/retry_context.go:93]: This was an existing bug, it should not be enabled by default

File: router/internal/retrytransport/retry_transport.go [hunk 3]
```diff
@@ -51,6 +35,61 @@ type RetryHTTPTransport struct {
...
+		return 0, false
+	}
+
+	duration := parseRetryAfterHeader(logger, retryAfter)
+	if duration > maxDuration {
+		duration = maxDuration
+	}
+
+	return duration, duration > 0
+}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/internal/retrytransport/retry_transport.go:91]: There is a default set in the config.go file, so it's fine. If it's 0 that means the user set it explicitly.

File: router/core/retry_expression_builder.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/wundergraph/cosmo/router/internal/expr"
+	"github.com/wundergraph/cosmo/router/internal/retrytransport"
+	"go.uber.org/zap"
+)
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/core/retry_expression_builder.go]: This is fine for now, we can add this later if a customer complains, the customer also can then modify the expression to handle this.

File: router-tests/go.mod [hunk 5]
```diff
@@ -70,7 +70,7 @@ require (
 	github.com/docker/docker-credential-helpers v0.9.3 // indirect
 	github.com/docker/go-units v0.5.0 // indirect
 	github.com/dustin/go-humanize v1.0.1 // indirect
-	github.com/expr-lang/expr v1.17.3 // indirect
+	github.com/expr-lang/expr v1.17.6 // indirect
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router-tests/go.mod:73]: Will update after PR is merged.

File: router/internal/retrytransport/retry_transport_test.go [hunk 6]
```diff
@@ -337,3 +449,693 @@ func createTestLogger(t *testing.T) (*bytes.Buffer, *zap.Logger) {
...
+			maxDuration := defaultMaxDuration
+			if tt.maxDuration > 0 {
+				maxDuration = tt.maxDuration
+			}
+			dur, use := shouldUseRetryAfter(zap.NewNop(), tt.resp, maxDuration)
+			assert.Equal(t, tt.expectedDur, dur)
+			assert.Equal(t, tt.expectedUse, use)
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/internal/retrytransport/retry_transport_test.go:1141]: This is not applicable since we did not clamp.

File: router/pkg/config/config.go [hunk 7]
```diff
@@ -234,6 +234,7 @@ type BackoffJitterRetry struct {
 	MaxAttempts int           `yaml:"max_attempts" envDefault:"5"`
 	MaxDuration time.Duration `yaml:"max_duration" envDefault:"10s"`
 	Interval    time.Duration `yaml:"interval" envDefault:"3s"`
+	Expression  string        `yaml:"expression,omitempty" env:"RETRY_EXPRESSION" envDefault:"IsRetryableStatusCode() || IsConnectionError() || IsTimeout()"`
```
Comment (Noroth) [inline router/pkg/config/config.go:237]: Why do we have an `env` variable for Expression and Enabled but not for the other properties?

File: router/core/retry_expression_builder.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+const DefaultRetryExpression = "IsRetryableStatusCode() || IsConnectionError() || IsTimeout()"
```
Comment (Noroth) [inline router/core/retry_expression_builder.go]: I think this doesn't have to be exported, or could be moved into the package `retrytransport`

File: router/core/retry_expression_builder.go [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	// Use default expression if empty string is passed
+	expression := retryOpts.Expression
+	if expression == "" {
+		expression = DefaultRetryExpression
+	}
+
+	// Create the retry expression manager
+	manager, err := expr.NewRetryExpressionManager(expression)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile retry expression: %w", err)
```
Comment (Noroth) [inline router/core/retry_expression_builder.go]: NewRetryExpressionManager already has a `failed to compile` errors message. This will create a duplicate error message

File: router/pkg/config/config.go (hunk 7 above)
Comment (SkArchon) [reply router/pkg/config/config.go:237]: Added to others

File: router/core/retry_expression_builder.go (hunk 9 above)
Comment (SkArchon) [reply router/core/retry_expression_builder.go]: Done

File: router/core/retry_expression_builder.go (hunk 8 above)
Comment (SkArchon) [reply router/core/retry_expression_builder.go]: Done

File: router/pkg/config/config.go (hunk 7 above)
Comment (Noroth) [reply router/pkg/config/config.go:237]: Why was algorithm left out?

Comment (SkArchon) [reply router/pkg/config/config.go:237]: It's technically a value that's not used anywhere at the moment, would prefer to discourage folks from using it.

File: router/core/supervisor_instance.go [hunk 10]
```diff
@@ -195,9 +195,12 @@ func optionsFromResources(logger *zap.Logger, config *config.Config) []Option {
 		WithSubgraphCircuitBreakerOptions(NewSubgraphCircuitBreakerOptions(config.TrafficShaping)),
 		WithSubgraphRetryOptions(
 			config.TrafficShaping.All.BackoffJitterRetry.Enabled,
+			config.TrafficShaping.All.BackoffJitterRetry.Algorithm,
 			config.TrafficShaping.All.BackoffJitterRetry.MaxAttempts,
 			config.TrafficShaping.All.BackoffJitterRetry.MaxDuration,
 			config.TrafficShaping.All.BackoffJitterRetry.Interval,
+			config.TrafficShaping.All.BackoffJitterRetry.Expression,
+			nil,
 		),
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/core/supervisor_instance.go:204]: This is fine because we specify env default values, which works unless its overriden by the user

File: router/core/router.go [hunk 11]
```diff
@@ -1752,13 +1752,25 @@ func WithSubgraphCircuitBreakerOptions(opts *SubgraphCircuitBreakerOptions) Opti
...
 			MaxRetryCount: maxRetryCount,
 			MaxDuration:   retryMaxDuration,
 			Interval:      retryInterval,
+			Expression:    expression,
+
+			// Test case overrides
+			OnRetry: onRetryFunc,
 		}
 	}
 }
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/core/router.go:1776]: Again this is fine, user will have to explicitly override it

File: router-tests/structured_logging_test.go [hunk 12]
```diff
@@ -709,7 +709,7 @@ func TestFlakyAccessLogs(t *testing.T) {
 					EnableSingleFlight:     true,
 					MaxConcurrentResolvers: 1,
 				}),
-				core.WithSubgraphRetryOptions(false, 0, 0, 0),
+				core.WithSubgraphRetryOptions(false, "", 0, 0, 0, "", nil),
 			},
```
[the remark this answers is not in the document]
Comment (SkArchon) [reply router-tests/structured_logging_test.go:713]: This is fine as we shouldn't validate when retries are disabled there

File: router/core/router.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (SkArchon) [reply router/core/router.go:1776]: Actually fixed

File: router/pkg/config/config.go (hunk 7 above)
Comment (SkArchon) [reply router/pkg/config/config.go:237]: Done, Updated

**Review by Noroth — APPROVED** [review]
